deadline_/social/models/managers/newsfeed_item.py

```diff
@@ -0,0 +1,87 @@
+"""
+The Manager for the NewsfeedItem model, which handles creation of multiple type of NewsfeedItems
+"""
+from django.db.models import Manager
+
+from accounts.models import User
+from challenges.models import UserSubcategoryProficiency, Submission, Challenge
+from social.constants import (
+    NW_ITEM_SUBCATEGORY_BADGE_POST, NW_ITEM_SHARE_POST,
+    NW_ITEM_SUBMISSION_LINK_POST, NW_ITEM_CHALLENGE_LINK_POST, NW_ITEM_CHALLENGE_COMPLETION_POST
+)
+
+
+class NewsfeedItemManager(Manager):
+    """
+    Use a custom NewsfeedItem manager for specific type of NewsfeedItem creation
+    """
+
+    def create_subcategory_badge_post(self, user_subcat_prof: UserSubcategoryProficiency):
+        """
+        Creates a SubcategoryBadgePost
+
+        A SubcategoryBadgePost is a NewsfeedItem which has the following fields as content:
+            proficiency_name: - the name of the proficiency (or badge) the user has attained
+            subcategory_name: - the name of the subcategory this badge is for
+            subcategory_id:   - the id of the subcategory
+
+        ex: Stanislav just earned the Master badge for Graph Algorithms!
+        """
+        return self.create(author_id=user_subcat_prof.user_id, type=NW_ITEM_SUBCATEGORY_BADGE_POST,
+                           content={
+                               'proficiency_name': user_subcat_prof.proficiency.name,
+                               'subcategory_name': user_subcat_prof.subcategory.name,
+                               'subcategory_id': user_subcat_prof.subcategory.id
+                           })
+
+    def create_share_post(self, shared_item: 'NewsfeedItem', author: User):
+        """
+        Creates a 'share' of a NewsfeedItem or in other words, a NewsfeedItem that points to another
+        """
+        # TODO: Add validation for not creating a share of a share
```

Comment:
Not sure if this is necessary. For example, FB allow that.

Reply:
I'm not actually sure Facebook allows infinite sharing of shares. Why allow to have such a long chain of shares when all it does is move the content away from the main post that's being shared?

Also, shares as they are right now do not hold any information inside them - only the ID of the shared item. This is because we want the client (front-end) to query for it himself. If we kept the shared item's information in it, if the original item ever got updated we'd either have inconsistencies or we'd have to update the value on all of the SHARES that share that item specifically.
As we currently only hold the ID of the shared item, imagine how many HTTP calls for a NWItem you'd need to make if an item had been shared 10 times.

Reply:
You might be right, however let's imagine the following scenario:

Someone shares a submission. Then at some future point he decides to re-share the same thing (probably it is something he's proud of etc.).  What is the expected behavior in this case?

Perhaps it will be good if we allow say up to 3 (or something like that) shares of the original one?

What do you think?

Reply:
You're confusing things.
A `share` of a submission is not the kind of `share` this is. Let's not call it a share at all, it's a submission post.

A `share` is a share of an already created post. In the scenario you gave, he would not have a share, he would have a submission post. At some future point, he will create a `share` of that submission post.
Imagine in some further future point, he sees this `share` he made. At this point, he is forbidden to share the `share` again. He is going to need to `share` the original submission post again.
This is how I imagine things

Reply:
Got it, thanks for the clarification. :)